src/ucp/tag/offload/rndv.c

```diff
@@ -0,0 +1,303 @@
+        return UCS_OK;
+    }
+
+    status = ucp_ep_resolve_remote_id(req->send.ep, rpriv->lane);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = ucp_proto_request_zcopy_init(
+                 req, rpriv->md_map, ucp_tag_rndv_offload_completion,
+                 UCT_MD_MEM_ACCESS_RMA | UCT_MD_MEM_FLAG_HIDE_ERRORS,
+                 UCP_DT_MASK_ALL);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucp_send_request_id_alloc(req);
+    req->flags |= UCP_REQUEST_FLAG_PROTO_INITIALIZED;
+
+    return UCS_OK;
+}
+
+UCS_PROFILE_FUNC(ucs_status_t, ucp_tag_rndv_offload_proto_progress, (self),
+                 uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_ep_h ep        = req->send.ep;
+    size_t max_iov     = ucp_ep_config(ep)->tag.offload.max_rndv_iov;
+    uct_iov_t *iov     = ucs_alloca(max_iov * sizeof(uct_iov_t));
+    const ucp_proto_rndv_ctrl_priv_t *rpriv;
+    ucp_tag_offload_unexp_rndv_hdr_t rndv_hdr;
+    ucp_datatype_iter_t next_iter;
+    ucp_lane_index_t lane;
+    ucs_status_t status;
+    void *rndv_op;
+    size_t iov_cnt;
+
+    status = UCS_PROFILE_CALL(ucp_tag_rndv_offload_request_init, req);
+    if (status != UCS_OK) {
+        ucp_proto_request_abort(req, status);
+        return UCS_OK;
+    }
+
+    rpriv             = req->send.proto_config->priv;
+    lane              = rpriv->lane;
+    rndv_hdr.ep_id    = ucp_send_request_get_ep_remote_id(req);
+    rndv_hdr.req_id   = ucp_send_request_get_id(req);
+    rndv_hdr.md_index = ucp_ep_md_index(ep, lane);
+
+    ucs_assert(!ucp_ep_use_indirect_id(req->send.ep));
+
+    iov_cnt = ucp_datatype_iter_next_iov(&req->send.state.dt_iter, SIZE_MAX,
+                                         rndv_hdr.md_index,
+                                         UCS_BIT(UCP_DATATYPE_CONTIG),
+                                         &next_iter, iov, max_iov);
+
+    rndv_op = uct_ep_tag_rndv_zcopy(ep->uct_eps[lane], req->send.msg_proto.tag,
+                                    &rndv_hdr, sizeof(rndv_hdr), iov, iov_cnt,
+                                    0, &req->send.state.uct_comp);
+
+    if (ucs_unlikely(UCS_PTR_IS_ERR(rndv_op))) {
+        status = UCS_PTR_STATUS(rndv_op);
+        return ucp_proto_single_status_handle(req, 1, NULL, lane, status);
+    }
+
+    req->flags                   |= UCP_REQUEST_FLAG_OFFLOADED;
+    req->send.tag_offload.rndv_op = rndv_op;
+
+    return UCS_OK;
+}
+
+static void
+ucp_tag_rndv_offload_proto_query(const ucp_proto_query_params_t *params,
+                                 ucp_proto_query_attr_t *attr)
+{
+    UCS_STRING_BUFFER_FIXED(strb, attr->config, sizeof(attr->config));
+    const ucp_proto_rndv_ctrl_priv_t *rpriv = params->priv;
+    ucp_context_h context                   = params->worker->context;
+    const ucp_ep_config_key_lane_t *ep_lane_cfg;
+    const ucp_tl_resource_desc_t *rsc;
+
+    ucs_assert(rpriv->lane < UCP_MAX_LANES);
```

Comment:
seems redundant

Reply:
it's needed to avoid coverity error.
I think [here](https://github.com/openucx/ucx/blob/master/src/ucp/proto/proto_common.c#L84) it is done for the same reason

Reply:
ok
maybe unite to helper func